### PR TITLE
verify command line options

### PR DESCRIPTION
**Pull request by rfomin**
Inspired by this issue: https://github.com/coelckers/prboom-plus/issues/515

I'm using `docgen` to generate a list of command line options to check.

### CODE REVIEW

File: src/m_argv.c [hunk 1]
```diff
@@ -69,6 +69,70 @@ boolean M_ParmExists(const char *check)
...
+void M_CheckCommandLine(void)
+{
+  int i;
+
+  for (i = 1; i < myargc; i++)
+  {
+    if (CheckParamsWithArgs(i))
+    {
+      // -file and -warp may have multiple arguments
+      if (!strcasecmp(myargv[i], "-file") ||
```
Comment (fabiangreffrath) [inline src/m_argv.c:116]: Does `-deh` also take multiple args?

Comment (rfomin) [reply src/m_argv.c]: `-warp` can take one or two arguments, depending on the version. Not sure if we should check everything.

Comment (fabiangreffrath) [reply src/m_argv.c]: Hm, "do it right or don't do it". 😁 Honestly, I'd expect command line checking code to detect that `-warp 1 2 3 4 5` cannot be right.

Comment (rfomin) [reply src/m_argv.c]: > Hm, "do it right or don't do it".

Yes, that's the question 😄 How do you like this approach, is it worth finishing it?
There is also `-recordfrom` which only takes two arguments.

Comment (rfomin) [reply src/m_argv.c]: BTW with `docgen` we can also generate code for `--help` options list, or maybe even `help <option>` messages.

Comment (fabiangreffrath) [reply src/m_argv.c]: We will also need to check the validity of the parameter passed to e.g. the `-complevel` argument, else this whole checking procedure won't make too much sense.

**Comment by kraflab** [comment]
https://github.com/kraflab/dsda-doom/commit/74a25a5bcd023ba1316fd5f5f10d79e88d379b23 maybe some food for thought, an approach in dsda-doom (see this as a more important precursor, but checking if they are used is easy to add on top).

File: src/m_argv.c (hunk 1 above)
Comment (rfomin) [reply src/m_argv.c]: I think we should check the validity of `-complevel` and other arguments in the game code, such as the `-file` argument if the path/file is not found. There is 36 options with arguments and most of them silently ignore wrong input.

**Comment by rfomin** [comment]
> [kraflab/dsda-doom@74a25a5](https://github.com/kraflab/dsda-doom/commit/74a25a5bcd023ba1316fd5f5f10d79e88d379b23) maybe some food for thought, an approach in dsda-doom (I see this as a more important precursor, but checking if they are used is easy to add on top).

Interesting, thanks. Looks like PrBoom+'s complevel is the only option that allows a negative integer as an argument? I want to use the `-` character as a delimiter.

**Comment by kraflab** [comment]
> Looks like PrBoom+'s complevel is the only option that allows a negative integer as an argument?

`-spechit` would "work" with a negative number I think, although I don't really know what the implication would be. It's an address offset to use for certain overflow problems in some specific demos :^)

`-setmem` also works with negative numbers. It's another overrun thing that I haven't looked into :^)

`-skipsec` takes negative numbers to skip from the end of the demo instead of from the start

There may be other stuff that takes `-1` for a default, I have no idea yet (but I guess I will have a full list when my refactoring is done...)

**Comment by rfomin** [comment]
> `-skipsec` takes negative numbers to skip from the end of the demo instead of from the start

Ha, I didn't know that and implemented it differently.

As for memory overflows, I think they are only positive, at least in the Choco code we copied, it is `unsigned`.

File: src/d_main.c [hunk 2]
```diff
@@ -2148,9 +2162,16 @@ void D_DoomMain(void)
...
-      autostart = true;
+      if (M_ParmStrToInt(myargv[p+1], &startepisode) &&
+          startepisode >= 1 && startepisode <= 4)
+      {
+        startmap = 1;
+        autostart = true;
+      }
+     else
+      {
+        I_Error("Wrong -episode parameter '%s', should be 1-4", myargv[p+1]);
```
Comment (kraflab) [inline src/d_main.c:2173]: Can't we have any episode from 0 to 9 now?

Comment (rfomin) [reply src/d_main.c:2173]: Right, in Woof we even support episodes 0-99.
@fabiangreffrath BTW `-warp 9999` doesn't currently work, only `-warp 99 99`. Not sure how to validate `-warp` 😄

Comment (fabiangreffrath) [reply src/d_main.c:2173]: I think it's okay as it is. No algorithm can know if `-warp 999` means to warp to E99M9 or E9M99. 😉

**Comment by fabiangreffrath** [comment]
> From my testing `-skill 0` doesn't work at all, all monsters and items still spawn.

This should emulate Vanilla behaviour:
```patch
--- a/src/p_mobj.c
+++ b/src/p_mobj.c
@@ -1209,10 +1209,11 @@ void P_SpawnMapThing (mapthing_t* mthing)
     return;

   // killough 11/98: simplify
-  if (gameskill == sk_baby || gameskill == sk_easy ?
+  if (gameskill == sk_none ||
+      (gameskill == sk_baby || gameskill == sk_easy ?
       !(mthing->options & MTF_EASY) :
       gameskill == sk_hard || gameskill == sk_nightmare ?
-      !(mthing->options & MTF_HARD) : !(mthing->options & MTF_NORMAL))
+      !(mthing->options & MTF_HARD) : !(mthing->options & MTF_NORMAL)))
     return;

   // [crispy] support MUSINFO lump (dynamic music changing)

```

**Comment by fabiangreffrath** [comment]
> This should emulate Vanilla behaviour:

Though, even Pr+ doesn't emulate this. But then the parameter help string needs to be adjusted.

**Comment by rfomin** [comment]
> Though, even Pr+ doesn't emulate this. But then the parameter help string needs to be adjusted.

We need to emulate it as we support Choco netgame so there is a possibility of desync. Interesting that `skill 0` doesn't work in demos.

**Comment by kraflab** [comment]
https://github.com/kraflab/dsda-doom/pull/116 if you're curious how that went🥵 
+1526 -936, slightly intrusive change :^)
Refactoring the use of 118 arguments is a real test of willpower 🤠 
But, never seeing `p < myargc - 1` in the code base is 💆

**Comment by rfomin** [comment]
> +1526 -936, slightly intrusive change :^)
> Refactoring the use of 118 arguments is a real test of willpower 🤠
> But, never seeing `p < myargc - 1` in the code base is 💆

I think Woof's motto is to keep as much code as possible, so I take the simpler approach 😄 We have 82 options, but there are a few undocumented secret ones.
Not sure if people will be happy with this change, it might break some launchers/.bat files.

File: man/docgen [hunk 3]
```diff
@@ -275,6 +302,21 @@ class Parameter:
...
 
+    def help_output(self, indent):
+        result = self.name
+
+        if self.args:
+            result += " " + self.args
+
+        result += " " * (indent - len(result))
+
+        result += textwrap.fill(self.text, width = (90 - indent),
```
Comment (rfomin) [inline man/docgen:313]: @fabiangreffrath Should we wrap help text for console output? Maybe we should write shorter help lines instead 😉

Comment (fabiangreffrath) [reply man/docgen:313]: Help text should fit into one console line ideally. Do you have an example of a help text line that doesn't fit?

Comment (rfomin) [reply man/docgen:313]: Oops missed your answer.
This is about a response to the `-help` command. I just generate our CMDLINE documentation, but with selected categories and parameters. So we have a lot of long lines here with multiple sentences.: [help.txt](https://github.com/fabiangreffrath/woof/files/9102301/help.txt)

File: src/i_system.c [hunk 4]
```diff
@@ -304,7 +304,13 @@ boolean I_GetMemoryValue(unsigned int offset, void *value, int size)
                         break;
                     }
 
-                    M_StrToInt(myargv[p], &val);
+                    if (!M_StrToInt(myargv[p], &val))
+                    {
+                        I_Error("Invalid parameter '%s' for -setmem, "
+                                "valid values are dos622, dos71, dosbox "
+                                "or memory offset.", myargv[p]);
```
Comment (kraflab) [inline src/i_system.c:311]: It's not an offset is it? I think it's up to 10 bytes of data.

Comment (rfomin) [reply src/i_system.c]: You're right, I'll correct it.

File: man/docgen [hunk 5]
```diff
@@ -531,6 +576,34 @@ def completion_output(targets, substs, template_file):
...
+    for t in targets:
+        content += t.carray_output(True)
+
+    content += "};\n\n#define HELP_STRING \""
+
+    c = dict(categories)
+
+    for t in targets:
+        # no video and obscure category
+        if t != c["video"] and t != c["obscure"]:
```
Comment (fabiangreffrath) [inline man/docgen:597]: Why no video?

File: src/d_main.c [hunk 6]
```diff
@@ -2127,30 +2150,49 @@ void D_DoomMain(void)
   // @category game
   // @arg <skill>
   // @vanilla
+  // @help
   //
-  // Set the game skill, 1-5 (1: easiest, 5: hardest).  A skill of
-  // 0 disables all monsters.
+  // Set the game skill, 1-5 (1: easiest, 5: hardest). A skill of 0 disables all
```
Comment (fabiangreffrath) [inline src/d_main.c:2155]: ...only in Vanilla complevel.

File: src/d_main.c [hunk 7]
```diff
@@ -2127,30 +2150,49 @@ void D_DoomMain(void)
...
+  // Start playing on episode n (0-99)
   //
 
   if ((p = M_CheckParm ("-episode")) && p < myargc-1)
     {
-      startepisode = myargv[p+1][0]-'0';
-      startmap = 1;
-      autostart = true;
+      startepisode = M_ParmArgToInt(p);
+      if (startepisode >= 0 && startepisode <= 99)
```
Comment (fabiangreffrath) [inline src/d_main.c:2186]: We are still unconditionally restricting episode to be > 0:

https://github.com/fabiangreffrath/woof/blob/d8a068c97b3e98b918d31360ad84c5b22f43c249/src/g_game.c#L2982-L2983

File: man/docgen (hunk 5 above)
Comment (rfomin) [reply man/docgen:597]: I don't think we have a video option important enough to include in the `-help` output. However, I'm not sure.

Comment (fabiangreffrath) [reply man/docgen:597]: Okay, the most important options already have their switches in the menu.

**Comment by fabiangreffrath** [comment]
So, is this ready for merging now?

**Comment by rfomin** [comment]
Yes, it's ready.

**Pull request merged**